# Importing libraries and dependencies

In [14]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!pip install folium
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


# Import neighbourhoods from previous notebooks

Credentials for reading/writing local assets on IBM Cloud Watson Studio

In [15]:
# The code was removed by Watson Studio for sharing.

Read file saved from previous notebook - scraped data from Wikipedia with longitude and latitude

In [17]:
# The code was removed by Watson Studio for sharing.

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

# How many unique Boroughs do we have? 

First lets see how many boroughs we have

In [30]:
unique_boroughs = neigh_df['Borough'].unique()
unique_borough_count = len(unique_boroughs)
unique_borough_count

10

Print all of the borough names

In [33]:
for i in range (0, unique_borough_count):
    print(unique_boroughs[i], "\n")

North York 

Downtown Toronto 

Etobicoke 

Scarborough 

East York 

York 

East Toronto 

West Toronto 

Central Toronto 

Mississauga 



We observe that 4 of the 10 boroughs are in Toronto

# Restrict our dataset to those borough which contian "Toronto" in their name

First identify which rows satisfy the condition that their Borough name contains Toronto

In [42]:
condition = neigh_df['Borough'].apply(lambda x: "Toronto" in x)

Copy the rows that satisfy this condition into a new dataframe

In [44]:
toronto_df = neigh_df[condition]

Reset indices

In [47]:
toronto_df.reset_index(drop=True, inplace=True)

In [48]:
toronto_df

Postal Code           Borough  \
0          M5A  Downtown Toronto   
1          M7A  Downtown Toronto   
2          M5B  Downtown Toronto   
3          M5C  Downtown Toronto   
4          M4E      East Toronto   
5          M5E  Downtown Toronto   
6          M5G  Downtown Toronto   
7          M6G  Downtown Toronto   
8          M5H  Downtown Toronto   
9          M6H      West Toronto   
10         M5J  Downtown Toronto   
11         M6J      West Toronto   
12         M4K      East Toronto   
13         M5K  Downtown Toronto   
14         M6K      West Toronto   
15         M4L      East Toronto   
16         M5L  Downtown Toronto   
17         M4M      East Toronto   
18         M4N   Central Toronto   
19         M5N   Central Toronto   
20         M4P   Central Toronto   
21         M5P   Central Toronto   
22         M6P      West Toronto   
23         M4R   Central Toronto   
24         M5R   Central Toronto   
25         M6R      West Toronto   
26         M4S   Central Toronto   
27         M5S  Downtown Toronto   
28         M6S      West Toronto   
29         M4T   Central Toronto   
30         M5T  Downtown Toronto   
31         M4V   Central Toronto   
32         M5V  Downtown Toronto   
33         M4W  Downtown Toronto   
34         M5W  Downtown Toronto   
35         M4X  Downtown Toronto   
36         M5X  Downtown Toronto   
37         M4Y  Downtown Toronto   
38         M7Y      East Toronto   

                                        Neighbourhood   Latitude  Longitude  
0                           Regent Park, Harbourfront  43.654260 -79.360636  
1         Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
2                            Garden District, Ryerson  43.657162 -79.378937  
3                                      St. James Town  43.651494 -79.375418  
4                                         The Beaches  43.676357 -79.293031  
5                                         Berczy Park  43.644771 -79.373306  
6                                  Central Bay Street  43.657952 -79.387383  
7                                            Christie  43.669542 -79.422564  
8                            Richmond, Adelaide, King  43.650571 -79.384568  
9                        Dufferin, Dovercourt Village  43.669005 -79.442259  
10  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
11                           Little Portugal, Trinity  43.647927 -79.419750  
12                       The Danforth West, Riverdale  43.679557 -79.352188  
13           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
14       Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191  
15                     India Bazaar, The Beaches West  43.668999 -79.315572  
16                     Commerce Court, Victoria Hotel  43.648198 -79.379817  
17                                    Studio District  43.659526 -79.340923  
18                                      Lawrence Park  43.728020 -79.388790  
19                                           Roselawn  43.711695 -79.416936  
20                                   Davisville North  43.712751 -79.390197  
21    Forest Hill North & West, Forest Hill Road Park  43.696948 -79.411307  
22                      High Park, The Junction South  43.661608 -79.464763  
23                  North Toronto West, Lawrence Park  43.715383 -79.405678  
24                The Annex, North Midtown, Yorkville  43.672710 -79.405678  
25                             Parkdale, Roncesvalles  43.648960 -79.456325  
26                                         Davisville  43.704324 -79.388790  
27                     University of Toronto, Harbord  43.662696 -79.400049  
28                                 Runnymede, Swansea  43.651571 -79.484450  
29                        Moore Park, Summerhill East  43.689574 -79.383160  
30          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049  
31  Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -

We have now restricted ourselves to the 39 postcodes within the 4 Toronto boroughs

# Get the latitude and longistude of Toronto

Use geopy library to get the latitude and longitude values of Toronto. We define an instance of the geocoder. We will name our agent <em>tor_explorer</em>.

In [50]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="tor_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


# Create a map of Toronto with Toronto neighbourhoods superimposed on top.

**In our current dataset we have that we have the longitudes and latitudes of our postcodes rather than the Neighbourhoods themselves. As such we will plot each the postcodes rather than the neighbourhoods.** 

The postcodes can contain multiple neighbourhoods, so on the map popup we provide the neighbourhoods associated with that postcode and the borough it is within.

In [65]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighbourhood, postcode in zip(toronto_df['Latitude'], toronto_df['Longitude'], toronto_df['Borough'], toronto_df['Neighbourhood'], toronto_df['Postal Code']):
    label = '{}; {}; {}'.format(postcode, neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

# Use the Foursquare API to exploed the neighbourhoods and segment them

Define Foursquare Credentials and Version

In [67]:
#@hidden cell
CLIENT_ID = 'XT2LDFOVZUGDEFYZU4A1PLM5X4UUP0204FYZ4W4H10LCOSII' # your Foursquare ID
CLIENT_SECRET = '0KMOFB3UXX4JDRU5A40EJMO3HMKHOKQHXGDV3G4ESSA0LYO0' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

# Get the top 100 venues for each postcode

Create a function that take the top 100 venues for each Postal Code in Toronto dataset

In [69]:
def getNearbyVenues(names, latitudes, longitudes, radius):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Postal Code', 
                  'Postal Code Latitude', 
                  'Postal Code Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Use the function defined to top venues in each Toronto postcode 

In [70]:
toronto_venues = getNearbyVenues(toronto_df['Postal Code'], toronto_df['Latitude'], toronto_df ['Longitude'], 500)

M5A
M7A
M5B
M5C
M4E
M5E
M5G
M6G
M5H
M6H
M5J
M6J
M4K
M5K
M6K
M4L
M5L
M4M
M4N
M5N
M4P
M5P
M6P
M4R
M5R
M6R
M4S
M5S
M6S
M4T
M5T
M4V
M5V
M4W
M5W
M4X
M5X
M4Y
M7Y


Check the size of the resulting dataframe and it's head

In [71]:
print(toronto_venues.shape)
toronto_venues.head()

(1624, 7)


Postal Code  Postal Code Latitude  Postal Code Longitude  \
0         M5A              43.65426             -79.360636   
1         M5A              43.65426             -79.360636   
2         M5A              43.65426             -79.360636   
3         M5A              43.65426             -79.360636   
4         M5A              43.65426             -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2  Cooper Koo Family YMCA       43.653249       -79.358008   
3     Body Blitz Spa East       43.654735       -79.359874   
4          Impact Kitchen       43.656369       -79.356980   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2  Distribution Center  
3                  Spa  
4           Restaurant

# Check how many venues were returned for each Postal Code

In [73]:
toronto_venues.groupby('Postal Code').count()

Postal Code Latitude  Postal Code Longitude  Venue  \
Postal Code                                                       
M4E                             4                      4      4   
M4K                            43                     43     43   
M4L                            19                     19     19   
M4M                            37                     37     37   
M4N                             3                      3      3   
M4P                             9                      9      9   
M4R                            18                     18     18   
M4S                            33                     33     33   
M4T                             2                      2      2   
M4V                            14                     14     14   
M4W                             4                      4      4   
M4X                            48                     48     48   
M4Y                            75                     75     75   
M5A                            44                     44     44   
M5B                           100                    100    100   
M5C                            85                     85     85   
M5E                            55                     55     55   
M5G                            68                     68     68   
M5H                           100                    100    100   
M5J                           100                    100    100   
M5K                           100                    100    100   
M5L                           100                    100    100   
M5N                             2                      2      2   
M5P                             4                      4      4   
M5R                            19                     19     19   
M5S                            34                     34     34   
M5T                            74                     74     74   
M5V                            16                     16     16   
M5W                            96                     96     96   
M5X                           100                    100    100   
M6G                            16                     16     16   
M6H                            13                     13     13   
M6J                            45                     45     45   
M6K                            23                     23     23   
M6P                            25                     25     25   
M6R                            14                     14     14   
M6S                            33                     33     33   
M7A                            33                     33     33   
M7Y                            16                     16     16   

             Venue Latitude  Venue Longitude  Venue Category  
Postal Code                                                   
M4E                       4                4               4  
M4K                      43               43              43  
M4L                      19               19              19  
M4M                      37               37              37  
M4N                       3                3               3  
M4P                       9                9               9  
M4R                      18               18              18  
M4S                      33               33              33  
M4T                       2                2               2  
M4V                      14               14              14  
M4W                       4                4               4  
M4X                      48               48              48  
M4Y                      75               75              75  
M5A                      44               44              44  
M5B                     100              100             100  
M5C                      85               85              85  
M5E                      55               55              55  
M5G                      68               68              68  
M5H  

# Encode each venue by type across Toronto Postal Codes

Use One-hot encoding categorise each venue type into a distinct type

In [74]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Postal Code'] = toronto_venues['Postal Code'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Postal Code  Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0         M5A                  0        0                   0             0   
1         M5A                  0        0                   0             0   
2         M5A                  0        0                   0             0   
3         M5A                  0        0                   0             0   
4         M5A                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  \
0                 0                   0              0          0           0   
1                 0                   0              0          0           0   
2                 0                   0              0          0           0   
3                 0                   0              0          0           0   
4                 0                   0              0          0           0   

   Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0           0       1     0    0                 0                   0      0   
1           0       0     0    0                 0                   0      0   
2           0       0     0    0                 0                   0      0   
3           0       0     0    0                 0                   0      0   
4           0       0     0    0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bistro  \
0                0         0           0                   0       0   
1                0         0           0                   0       0   
2                0         0           0                   0       0   
3                0         0           0                   0       0   
4                0         0           0                   0       0   

   Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  \
0              0          0         0                     0               0   
1              0          0         0                     0               0   
2              0          0         0                     0               0   
3              0          0         0                     0               0   
4              0          0         0                     0               0   

   Brewery  Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0        0                0         0             0              0         0   
1        0                0         0             0              0         0   
2        0                0         0             0              0         0   
3        0                0         0             0              0         0   
4        0                0         0             0              0         0   

   Butcher  Café  Cajun / Creole Restaurant  Camera Store  Candy Store  \
0        0     0                          0             0            0   
1        0     0                          0             0            0   
2        0     0            

And let's examine the new dataframe size.

In [76]:
toronto_onehot.shape

(1624, 238)

Note that we have kept all 1624 venues in all the Toronto postal codes

Also we see that there are **238 different venue types in Toronto**

# Group over Postal Codes using the mean to find the average number of venues of that type within a Postal Code

Let's group rows by Postal code and by taking the mean of the frequency of occurrence of each category


In [77]:
toronto_grouped = toronto_onehot.groupby('Postal Code').mean().reset_index()
toronto_grouped

Postal Code  Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0          M4E           0.000000   0.0000              0.0000        0.0000   
1          M4K           0.000000   0.0000              0.0000        0.0000   
2          M4L           0.000000   0.0000              0.0000        0.0000   
3          M4M           0.000000   0.0000              0.0000        0.0000   
4          M4N           0.000000   0.0000              0.0000        0.0000   
5          M4P           0.000000   0.0000              0.0000        0.0000   
6          M4R           0.000000   0.0000              0.0000        0.0000   
7          M4S           0.000000   0.0000              0.0000        0.0000   
8          M4T           0.000000   0.0000              0.0000        0.0000   
9          M4V           0.000000   0.0000              0.0000        0.0000   
10         M4W           0.000000   0.0000              0.0000        0.0000   
11         M4X           0.000000   0.0000              0.0000        0.0000   
12         M4Y           0.013333   0.0000              0.0000        0.0000   
13         M5A           0.000000   0.0000              0.0000        0.0000   
14         M5B           0.000000   0.0000              0.0000        0.0000   
15         M5C           0.000000   0.0000              0.0000        0.0000   
16         M5E           0.000000   0.0000              0.0000        0.0000   
17         M5G           0.000000   0.0000              0.0000        0.0000   
18         M5H           0.000000   0.0000              0.0000        0.0000   
19         M5J           0.000000   0.0000              0.0000        0.0000   
20         M5K           0.000000   0.0000              0.0000        0.0000   
21         M5L           0.000000   0.0000              0.0000        0.0000   
22         M5N           0.000000   0.0000              0.0000        0.0000   
23         M5P           0.000000   0.0000              0.0000        0.0000   
24         M5R           0.000000   0.0000              0.0000        0.0000   
25         M5S           0.000000   0.0000              0.0000        0.0000   
26         M5T           0.000000   0.0000              0.0000        0.0000   
27         M5V           0.000000   0.0625              0.0625        0.0625   
28         M5W           0.000000   0.0000              0.0000        0.0000   
29         M5X           0.000000   0.0000              0.0000        0.0000   
30         M6G           0.000000   0.0000              0.0000        0.0000   
31         M6H           0.000000   0.0000              0.0000        0.0000   
32         M6J           0.000000   0.0000              0.0000        0.0000   
33         M6K           0.000000   0.0000              0.0000        0.0000   
34         M6P           0.000000   0.0000              0.0000        0.0000   
35         M6R           0.000000   0.0000              0.0000        0.0000   
36         M6S           0.000000   0.0000              0.0000        0.0000   
37         M7A           0.000000   0.0000              0.0000        0.0000   
38         M7Y           0.000000   0.0000              0.0000        0.0000   

    Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0            0.000            0.000            0.0000             0.000000   
1            0.000            0.000            0.0000             0.023256   
2            0.000            0.000            0.0000             0.000000   
3            0.000            0.000            0.0000             0.054054   
4            0.000            0.000            0.0000             0.000000   
5            0.000            0.000            0.0000             0.000000   
6            0.000            0.000            0.0000             0.000000   
7            0.000            0.000            0.0000             0.000000   
8            0.000            0.000            0.0000             0.000000   
9            0.000    

# Let's print each Postal Code along with the top 5 most common venues

In [78]:
num_top_venues = 5

for post_code in toronto_grouped['Postal Code']:
    print("----"+post_code+"----")
    temp = toronto_grouped[toronto_grouped['Postal Code'] == post_code].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----M4E----
                venue  freq
0               Trail  0.25
1   Health Food Store  0.25
2        Neighborhood  0.25
3                 Pub  0.25
4  Mexican Restaurant  0.00


----M4K----
                    venue  freq
0        Greek Restaurant  0.16
1             Coffee Shop  0.09
2      Italian Restaurant  0.07
3  Furniture / Home Store  0.05
4               Bookstore  0.05


----M4L----
                  venue  freq
0                  Park  0.11
1           Coffee Shop  0.05
2               Brewery  0.05
3  Fast Food Restaurant  0.05
4     Fish & Chips Shop  0.05


----M4M----
                 venue  freq
0          Coffee Shop  0.08
1              Brewery  0.05
2                 Café  0.05
3  American Restaurant  0.05
4               Bakery  0.05


----M4N----
                 venue  freq
0                 Park  0.33
1             Bus Line  0.33
2          Swim School  0.33
3    Afghan Restaurant  0.00
4  Monument / Landmark  0.00


----M4P----
                  venue  freq


# Lets turn this into a dataframe with the top 10 venues

First, let's write a function to sort the venues in descending order.

In [184]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each Postal Code.

In [185]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Postal Code']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
post_code_venues_sorted = pd.DataFrame(columns=columns)
post_code_venues_sorted['Postal Code'] = toronto_grouped['Postal Code']

for ind in np.arange(toronto_grouped.shape[0]):
    post_code_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

post_code_venues_sorted.head()

Postal Code 1st Most Common Venue 2nd Most Common Venue  \
0         M4E                   Pub     Health Food Store   
1         M4K      Greek Restaurant           Coffee Shop   
2         M4L                  Park      Sushi Restaurant   
3         M4M           Coffee Shop   American Restaurant   
4         M4N                  Park              Bus Line   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                 Trail          Neighborhood           Yoga Studio   
1    Italian Restaurant            Restaurant        Ice Cream Shop   
2     Fish & Chips Shop            Steakhouse                   Pub   
3                Bakery               Brewery                  Café   
4           Swim School    Dim Sum Restaurant  Ethiopian Restaurant   

    6th Most Common Venue 7th Most Common Venue        8th Most Common Venue  \
0     Distribution Center    Dim Sum Restaurant                        Diner   
1  Furniture / Home Store             Bookstore                    Juice Bar   
2                 Brewery  Fast Food Restaurant           Italian Restaurant   
3               Gastropub           Yoga Studio                  Fish Market   
4             Escape Room     Electronics Store  Eastern European Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0        Discount Store       Doner Restaurant  
1                   Spa                    Pub  
2            Restaurant            Pizza Place  
3                  Park           Neighborhood  
4   Dumpling Restaurant             Donut Shop

# Cluster the Postal Codes

Run _k_-means to cluster the neighborhood into 4 clusters.

In [186]:
# set number of clusters
kclusters = 4

toronto_grouped_clustering = toronto_grouped.drop('Postal Code', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0, init='k-means++')
                
#Fit data to model
kmeans.fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 2, 0, 0, 0, 3, 0], dtype=int32)

Lets add the clustering labels to the Postal Codes with their top 10 venues

In [187]:
# add clustering labels
post_code_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

In [188]:
post_code_venues_sorted.head()

Cluster Labels Postal Code 1st Most Common Venue 2nd Most Common Venue  \
0               0         M4E                   Pub     Health Food Store   
1               0         M4K      Greek Restaurant           Coffee Shop   
2               0         M4L                  Park      Sushi Restaurant   
3               0         M4M           Coffee Shop   American Restaurant   
4               2         M4N                  Park              Bus Line   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                 Trail          Neighborhood           Yoga Studio   
1    Italian Restaurant            Restaurant        Ice Cream Shop   
2     Fish & Chips Shop            Steakhouse                   Pub   
3                Bakery               Brewery                  Café   
4           Swim School    Dim Sum Restaurant  Ethiopian Restaurant   

    6th Most Common Venue 7th Most Common Venue        8th Most Common Venue  \
0     Distribution Center    Dim Sum Restaurant                        Diner   
1  Furniture / Home Store             Bookstore                    Juice Bar   
2                 Brewery  Fast Food Restaurant           Italian Restaurant   
3               Gastropub           Yoga Studio                  Fish Market   
4             Escape Room     Electronics Store  Eastern European Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0        Discount Store       Doner Restaurant  
1                   Spa                    Pub  
2            Restaurant            Pizza Place  
3                  Park           Neighborhood  
4   Dumpling Restaurant             Donut Shop

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each Postal Code.

In [189]:
toronto_merged = toronto_df

# merge toronto_grouped with toronto_data to add latitude/longitude for each Postal Code
toronto_merged = toronto_merged.join(post_code_venues_sorted.set_index('Postal Code'), on='Postal Code')

toronto_merged.shape

(39, 16)

In [190]:
toronto_merged.head()

Postal Code           Borough                                Neighbourhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.654260 -79.360636               0           Coffee Shop   
1  43.662301 -79.389494               0           Coffee Shop   
2  43.657162 -79.378937               0        Clothing Store   
3  43.651494 -79.375418               0           Coffee Shop   
4  43.676357 -79.293031               0                   Pub   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                   Pub                Bakery                  Park   
1                  Bank              Beer Bar         Smoothie Shop   
2           Coffee Shop                  Café       Bubble Tea Shop   
3                  Café          Cocktail Bar            Restaurant   
4     Health Food Store                 Trail          Neighborhood   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0        Breakfast Spot                  Café               Theater   
1        Sandwich Place            Restaurant                  Café   
2   Japanese Restaurant        Cosmetics Shop                 Hotel   
3             Gastropub   American Restaurant              Beer Bar   
4           Yoga Studio   Distribution Center    Dim Sum Restaurant   

   8th Most Common Venue 9th Most Common Venue     10th Most Common Venue  
0            Yoga Studio           Event Space                 Restaurant  
1  Portuguese Restaurant    Chinese Restaurant                       Park  
2              Bookstore           Pizza Place  Middle Eastern Restaurant  
3                    Gym   Moroccan Restaurant           Department Store  
4                  Diner        Discount Store           Doner Restaurant

# How many Postal Codes are there in each cluster?

In [191]:
toronto_merged['Cluster Labels'].value_counts()

0    35
2     2
3     1
1     1
Name: Cluster Labels, dtype: int64

# Finally, let's visualize the resulting clusters

In [192]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, neighbourhood, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Postal Code'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + "\n" + neighbourhood + "\n" + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

# Examine clusters

Now, you can examine each cluster and determine the discriminating venue categories that distinguish each cluster.

### Cluster 1 - Places to Drink or Eat (Coffee Shops, Cafe, Pubs or Bakery Restaurent)

In [193]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[0] + list(range(5, toronto_merged.shape[1]))]]

Postal Code  Cluster Labels 1st Most Common Venue  \
0          M5A               0           Coffee Shop   
1          M7A               0           Coffee Shop   
2          M5B               0        Clothing Store   
3          M5C               0           Coffee Shop   
4          M4E               0                   Pub   
5          M5E               0           Coffee Shop   
6          M5G               0           Coffee Shop   
7          M6G               0         Grocery Store   
8          M5H               0           Coffee Shop   
9          M6H               0                Bakery   
10         M5J               0           Coffee Shop   
11         M6J               0                   Bar   
12         M4K               0      Greek Restaurant   
13         M5K               0           Coffee Shop   
14         M6K               0                  Café   
15         M4L               0                  Park   
16         M5L               0           Coffee Shop   
17         M4M               0           Coffee Shop   
20         M4P               0  Gym / Fitness Center   
21         M5P               0                  Park   
22         M6P               0                  Café   
23         M4R               0           Coffee Shop   
24         M5R               0        Sandwich Place   
25         M6R               0        Breakfast Spot   
26         M4S               0        Sandwich Place   
27         M5S               0                  Café   
28         M6S               0           Coffee Shop   
30         M5T               0                   Bar   
31         M4V               0           Coffee Shop   
32         M5V               0        Airport Lounge   
34         M5W               0           Coffee Shop   
35         M4X               0           Coffee Shop   
36         M5X               0           Coffee Shop   
37         M4Y               0           Coffee Shop   
38         M7Y               0  Gym / Fitness Center   

            2nd Most Common Venue 3rd Most Common Venue  \
0                             Pub                Bakery   
1                            Bank              Beer Bar   
2                     Coffee Shop                  Café   
3                            Café          Cocktail Bar   
4               Health Food Store                 Trail   
5                    Cocktail Bar    Seafood Restaurant   
6                            Café        Sandwich Place   
7                            Café                  Park   
8                            Café                 Hotel   
9                        Pharmacy         Grocery Store   
10                       Aquarium                 Hotel   
11                    Coffee Shop           Men's Store   
12                    Coffee Shop    Italian Restaurant   
13                          Hotel                  Café   
14                      Nightclub           Coffee Shop   
15               Sushi Restaurant     Fish & Chips Shop   
16                     Restaurant                 Hotel   
17            American Restaurant                Bakery   
20                 Sandwich Place                  Park   
21                  Jewelry Store                 Trail   
22                Thai Restaurant                   Bar   
23                 Clothing Store           Yoga Studio   
24                           Café           Coffee Shop   
25                      Gift Shop             Bookstore   
26                   Dessert Shop           Pizza Place   
27                      Bookstore                   Bar   
28                           Café      Sushi Restaurant   
30  Vegetarian / Vegan Restaurant           Coffee Shop   
31                    Pizza Place    Light Rail Station   
32                Airport Service              Boutique   
34             Italian Restaurant                 Hotel   
35                     Restaurant           Pizza Place   
36                           Café      

### Cluster 2 - Music Venue

In [194]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[0] + list(range(5, toronto_merged.shape[1]))]]

Postal Code  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
19         M5N               1           Music Venue                Garden   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
19           Yoga Studio          Dessert Shop  Ethiopian Restaurant   

   6th Most Common Venue 7th Most Common Venue        8th Most Common Venue  \
19           Escape Room     Electronics Store  Eastern European Restaurant   

   9th Most Common Venue 10th Most Common Venue  
19   Dumpling Restaurant             Donut Shop

This cluster includes postcodes with Music Venue

### Cluster 3 - Parks

In [195]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[0] + list(range(5, toronto_merged.shape[1]))]]

Postal Code  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
18         M4N               2                  Park              Bus Line   
33         M4W               2                  Park            Playground   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
18           Swim School    Dim Sum Restaurant  Ethiopian Restaurant   
33                 Trail      Department Store           Escape Room   

   6th Most Common Venue        7th Most Common Venue  \
18           Escape Room            Electronics Store   
33     Electronics Store  Eastern European Restaurant   

          8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
18  Eastern European Restaurant   Dumpling Restaurant             Donut Shop  
33          Dumpling Restaurant            Donut Shop       Doner Restaurant

This cluster includes postcodes with Parks

### Cluster 4 - Playgrounds

In [196]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[0] + list(range(5, toronto_merged.shape[1]))]]

Postal Code  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
29         M4T               3            Playground                 Trail   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
29           Yoga Studio      Department Store           Escape Room   

   6th Most Common Venue        7th Most Common Venue 8th Most Common Venue  \
29     Electronics Store  Eastern European Restaurant   Dumpling Restaurant   

   9th Most Common Venue 10th Most Common Venue  
29            Donut Shop       Doner Restaurant

This cluster includes postcodes with Playgrounds